core/src/main/java/com/linecorp/armeria/common/util/DefaultBlockingTaskExecutor.java

```diff
@@ -17,43 +17,114 @@
 package com.linecorp.armeria.common.util;
 
 import java.util.Collection;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 import java.util.concurrent.Callable;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.collect.ImmutableList;
 
 final class DefaultBlockingTaskExecutor implements BlockingTaskExecutor {
 
     private final ScheduledExecutorService delegate;
 
+    private final AtomicInteger taskCounter = new AtomicInteger(0);
+
     DefaultBlockingTaskExecutor(ScheduledExecutorService delegate) {
         this.delegate = delegate;
     }
 
     @Override
     public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit) {
-        return delegate.schedule(command, delay, unit);
+        boolean submitted = false;
+        taskCounter.incrementAndGet();
+        try {
+            final ScheduledFuture<?> future = delegate.schedule(() -> {
+                try {
+                    command.run();
+                } finally {
+                    taskCounter.decrementAndGet();
+                }
+            }, delay, unit);
+            submitted = true;
+            return future;
+        } finally {
+            if (!submitted) {
+                taskCounter.decrementAndGet();
+            }
+        }
     }
 
     @Override
     public <V> ScheduledFuture<V> schedule(Callable<V> callable, long delay, TimeUnit unit) {
-        return delegate.schedule(callable, delay, unit);
+        boolean submitted = false;
+        taskCounter.incrementAndGet();
+        try {
+            final ScheduledFuture<V> future = delegate.schedule(() -> {
+                try {
+                    return callable.call();
+                } finally {
+                    taskCounter.decrementAndGet();
+                }
+            }, delay, unit);
+            submitted = true;
+            return future;
+        } finally {
+            if (!submitted) {
+                taskCounter.decrementAndGet();
+            }
+        }
     }
 
     @Override
     public ScheduledFuture<?> scheduleAtFixedRate(Runnable command, long initialDelay, long period,
                                                   TimeUnit unit) {
-        return delegate.scheduleAtFixedRate(command, initialDelay, period, unit);
+        boolean submitted = false;
+        taskCounter.incrementAndGet();
+        try {
+            final ScheduledFuture<?> future = delegate.scheduleAtFixedRate(() -> {
+                try {
+                    command.run();
+                } finally {
+                    taskCounter.decrementAndGet();
```

Comment:
I think `taskCounter` will be decremented on each scheduled task invocation, but no increment occurs - this may not correctly reflect the number of pending tasks